test/components/project/list.spec.js

```diff
@@ -0,0 +1,124 @@
+    it('redirects an anonymous user to login', () =>
+      mockRoute('/')
+        .restoreSession(false)
+        .afterResponse(app => {
+          app.vm.$route.path.should.equal('/login');
+        }));
+
+    it('redirects the user back after login', () =>
+      mockRouteThroughLogin('/')
+        .respondWithData(() => testData.extendedProjects.createPast(1).sorted())
+        .respondWithData(() => testData.administrators.sorted())
+        .afterResponses(app => {
+          app.vm.$route.path.should.equal('/');
+        }));
+
+    for (const selector of ['.navbar-brand', '#navbar-projects-link']) {
+      it(`redirects user to project list upon a click on ${selector}`, () => {
+        mockLogin();
+        return mockRoute('/users')
+          .respondWithData(() => testData.administrators.sorted())
+          .complete()
+          .request(app => trigger.click(app, selector))
+          .respondWithData(() => testData.extendedProjects.createPast(1).sorted())
+          .respondWithData(() => testData.administrators.sorted())
+          .afterResponses(app => {
+            app.vm.$route.path.should.equal('/');
+          });
+      });
+    }
+  });
+
+  describe('after login', () => {
+    beforeEach(mockLogin);
+
+    describe('Right Now', () => {
+      it('shows counts', () =>
+        mockRoute('/')
+          .respondWithData(() => testData.extendedProjects.createPast(2).sorted())
+          .respondWithData(() => testData.administrators.sorted())
+          .afterResponses(app => {
+            const counts = app.find('.project-list-right-now-count');
+            counts.map(count => count.text().trim()).should.eql(['1', '2']);
+          }));
+
+      const targets = [
+        ['icon', '.project-list-right-now-icon-container'],
+        ['count', '.project-list-right-now-count a'],
+        ['description', '.project-list-right-now-description a']
+      ];
+      for (const [description, selector] of targets) {
+        it(`renders a link to /users for the users ${description}`, () =>
+          mockRoute('/')
+            .respondWithData(() => testData.extendedProjects.createPast(1).sorted())
+            .respondWithData(() => testData.administrators.sorted())
+            .afterResponses(app => {
+              app.first(selector).getAttribute('href').should.equal('#/users');
+            }));
+
+        it(`scrolls down the page upon a click on the projects ${description}`, () =>
+          mockRoute('/', { attachToDocument: true })
+            .respondWithData(() => testData.extendedProjects.createPast(1).sorted())
+            .respondWithData(() => testData.administrators.sorted())
+            .afterResponses(app => {
+              window.pageYOffset.should.equal(0);
+              return trigger.click(app.find(selector)[1]);
+            })
+            // Wait for the animation to complete.
+            .then(() => new Promise(resolve => {
+              setTimeout(resolve, 400);
```

Comment:
After thinking about it, I think this probably shouldn't be set to a smaller number: if it were, the animation could still be in progress during proceeding tests. However, I think we have other options if/when we want to address this: I think there are a couple of ways that we could make the animation complete in a shorter time during testing.

Reply:
my point here was that your assertion simply checks for a nonzero scrolltop, so if the animation has _begun at all_ it will pass.

Reply:
That makes sense to me, and it would be nice to make this test shorter. What I realized though was that since Karma uses a single document for all tests, if we initiate the scroll in this test but don't wait until the scroll completes, I think the document may still be scrolling when the next test starts.

Reply:
i think that's not a problem?

Reply:
I think you're right that it usually shouldn't be. The only problematic scenario that comes to mind is if one scroll-related test runs soon after another scroll-related test, since the scrolling within the first test may affect the second test. There are three scroll-related tests here, and when I try changing the timeout from `400` to `300`, the first test passes, but the other two end up failing.

One idea I had was that we could use `envify` (already one of our dependencies) to specify one animation duration for production and another for testing. Something like:

```js
const duration = process.env.NODE_ENV !== "test" ? 400 : 1;
$('html, body').animate({ scrollTop }, duration);
```

How does something like that sound?

Reply:
i think test should run the same way prod does. i think you could probably cancel all animations before every test, or before scrolling-related tests? but i guess this isn't worth this much time right now.